working/1426-extension-types/feature-specification-implements-rep.md

```diff
@@ -0,0 +1,268 @@
+# Inline Classes can Implement the Representation Type
+
+Author: eernst@google.com
+
+Status: Draft.
+
+This is an addendum to the [inline class feature specification][],
+proposing an additional feature: An inline class `V` may have its
+representation type `R` as a superinterface. This causes `V` to be a
+subtype of `R`, and it enables invocations of members of `R` on a receiver
+of type `V`.
+
+[inline class feature specification]: https://github.com/dart-lang/language/blob/main/accepted/future-releases/inline-classes/feature-specification.md
+
+## Change Log
+
+## Summary
+
+This is a proposal to add a new feature to inline classes: An inline class `V`
+may include one or more superinterfaces in its `implements` clause which are
```

Comment:
Cool stuff! ;-)

First, though, I think this choice between 'superinterface' and 'supertype' is irrelevant: The direct superinterfaces of a given declaration that introduces a type (including `inline class` declarations) are the types mentioned in certain clauses (in this case we only have `implements`, but with two different kinds of elements in them). The indirect superinterfaces are just derived from that. So with `implements T` where `T` is a non-inline type, `T` must be a supertype of the representation type, and the _role_ that it plays is that of a direct superinterface.

Next, I would say that non-interface types should be included already among the types that we can specify like this, but I added a few words to emphasize that we actually mean it when we say that it can be any supertype of the representation type, and then I also used the `MapEntry` example. Nice!

I hadn't thought about the syntax issue. I haven't done anything about it yet, and a typedef can certainly give the appropriate syntax for any type. Otherwise, I tend to think (again!) that we should use a separate syntactic construct in order to specify declared subtype relationships to non-inline types. Perhaps `reveals T1, .. Ts`.

In any case, it _does_ make a difference that we wish to enable `int?` and `void Function([String])` as non-inline superinterfaces of an inline class, but non-inline classes can't have any of those.

I think it will work with multiple function types and the redeclared `call` method.

I don't think we need anything special for members of records: They are getters, they will be required to have a combined member signature, and that's well defined already. A subtype of two record types will need to be so based on covariance, because different shapes have unrelated types (or it's a bottom type, but an inline class whose representation type is `Never` isn't very important anyway).

`dynamic`: It would be nice if we could make this mean "invoke the members declared in the inline class statically; invoke anything else dynamically (except that `Object` members are somewhat statically known, as usual)". It would certainly conflict with any other superinterface because `dynamic` will always be able to come up with a member signature which is in conflict with that other superinterface. However, I've made it an error for now.

`void`: As usual, I think this should be an error.

`Null`: Anything special here? The representation type is probably `Null`, too.

`Never`: Similar. Probably OK, because it won't happen, and if someone does it then it's useless, but benign.

`Function`: Probably benign and useless again.